[changelog] Validate the content before saving

vchrombie opened the pull request:
- [x] check if the content is in a valid format 

Fixes #17

vchrombie commented:
Hi @sduenas 
I have marked the PR for review. I have implemented two checks regarding the author and issue number fields

> * [x]  author field is mandatory and should be in the format `John Smith <jsmith@example.com>`
> * [x]  issue field is not mandatory (can be empty) but it should be an integer of present

Do you think the `issue` number should be mandatory? 
Also, should I implement a check for `notes` field?
```
notes: >
  The bug was making impossible to cast a spell on
  a magician.
```


Please let me know your comments.

sduenas commented on release_tools/changelog.py:
```diff
@@ -208,16 +211,36 @@ def create_changelog_entry_content(title, category, author=None, issue=None,
+        msg = "Author field should be of the format `John Smith <jsmith@example.com>`"
+        raise click.ClickException(msg)
+
+    if data['issue']:
+        try:
+            int(data['issue'])
+        except ValueError:
+            msg = "Issue number must be an integer"
+            raise click.ClickException(msg)
```
I don't think we need to check if it's an integer or not. For example, in Jira ids have text and numbers.

sduenas commented on release_tools/changelog.py:
```diff
@@ -208,16 +211,36 @@ def create_changelog_entry_content(title, category, author=None, issue=None,
+        msg = "File has an issue on line {} at position {}".format(pm.line, pm.column)
+        raise click.ClickException(msg)
+
+    if data['author'] and re.search(REGEX, data['author']):
+        pass
+    else:
+        msg = "Author field should be of the format `John Smith <jsmith@example.com>`"
+        raise click.ClickException(msg)
```
I don't think we need to check the format of the author. There might be people using nicknames only. Remember that the idea is that this tool can work for everybody or any python project.

sduenas commented on release_tools/changelog.py:
```diff
@@ -208,16 +211,36 @@ def create_changelog_entry_content(title, category, author=None, issue=None,
     else:
         content = stream
 
+    if not content:
+        msg = "Aborting due to empty entry content"
+        raise click.ClickException(msg)
+
+    try:
+        data = yaml.safe_load(content)
+    except yaml.YAMLError as exc:
+        pm = exc.problem_mark
+        msg = "File has an issue on line {} at position {}".format(pm.line, pm.column)
+        raise click.ClickException(msg)
```
All of this should be part of a different function to separate the logic.

vchrombie replied:
Yeah, makes much sense.

sduenas reviewed (changes requested):
I think we don't need to check the contents of most of the fields. We need to know whether is a valid yaml document or not (you already implemented that), check that the file is not empty.

sduenas commented:
> Hi @sduenas
> I have marked the PR for review. I have implemented two checks regarding the author and issue number fields
> 
> > * [x]   author field is mandatory and should be in the format `John Smith <jsmith@example.com>`
> > * [x]   issue field is not mandatory (can be empty) but it should be an integer of present
> 
> Do you think the `issue` number should be mandatory?
> Also, should I implement a check for `notes` field?
> 
> ```
> notes: >
>   The bug was making impossible to cast a spell on
>   a magician.
> ```
> 
> Please let me know your comments.



> Hi @sduenas
> I have marked the PR for review. I have implemented two checks regarding the author and issue number fields
> 
> > * [x]   author field is mandatory and should be in the format `John Smith <jsmith@example.com>`
> > * [x]   issue field is not mandatory (can be empty) but it should be an integer of present
> 
> Do you think the `issue` number should be mandatory?

I think all the fields should appear on the document but they can contain empty values.

> Also, should I implement a check for `notes` field?
> 
> ```
> notes: >
>   The bug was making impossible to cast a spell on
>   a magician.
> ```
> 

I think that makes sense so we avoid problems with the format. In YAML there are many ways to write multilne string but we can force to use this one.




> Please let me know your comments.

vchrombie commented:
Hi @sduenas, I have addressed your comments. I also added an option to edit the file incase if it has any mistakes. I felt aborting because of a typo, after typing the whole thing, might not be satisfactory. :slightly_frowning_face: 

It would look something like this
```
Error: File has an issue on line 3 at position 17
Do you want to edit it? [y/N]: y
Error: File has an issue on line 4 at position 9
Do you want to edit it? [y/N]: n
Error: Aborting due to wrong file format
```

Should I revert it? WDYT?

sduenas commented on release_tools/changelog.py:
```diff
@@ -211,13 +212,33 @@ def create_changelog_entry_content(title, category, author=None, issue=None,
+            return content
+        except yaml.YAMLError as exc:
+            pm = exc.problem_mark
+            click.echo("Error: File has an issue on line {} at position {}".format(pm.line, pm.column))
```
As there's not file created yet, maybe it's better to say `Invalid format on ...`

sduenas commented on release_tools/changelog.py:
```diff
@@ -211,13 +212,33 @@ def create_changelog_entry_content(title, category, author=None, issue=None,
+            click.echo("Error: File has an issue on line {} at position {}".format(pm.line, pm.column))
+
+            if not click.confirm("Do you want to edit it?"):
+                msg = "Aborting due to wrong file format"
```
Should this abort or just save the entry with the errors?

vchrombie replied:
If the user selected "N" (no), it would abort, doesn't save the file with errors.

vchrombie replied:
If the user selects "y" (yes), it would show the errors and open the editor. This process would continue either the user selects no or corrects all the errors.

vchrombie replied:
> Should this abort or just save the entry with the errors?

saving the entry might create erros later, so I implemented abort :thinking: 
WDYT?

sduenas commented on release_tools/changelog.py:
```diff
@@ -211,13 +212,33 @@ def create_changelog_entry_content(title, category, author=None, issue=None,
-    """Store the contents of an entry in a file."""
-
+def validate_changelog_content(content, run_editor=True):
     if not content:
```
This should also be checked after the user edits again the file when there's an error.

sduenas commented on release_tools/changelog.py:
```diff
@@ -211,13 +212,33 @@ def create_changelog_entry_content(title, category, author=None, issue=None,
         msg = "Aborting due to empty entry content"
         raise click.ClickException(msg)
 
+    while True:
```
I don't like this infinite loop here. I think it's better to exit the loop when the content is validated.

sduenas reviewed (changes requested):
Please check my comments.

sduenas commented:
> Hi @sduenas, I have addressed your comments. I also added an option to edit the file incase if it has any mistakes. I felt aborting because of a typo, after typing the whole thing, might not be satisfactory. slightly_frowning_face
> 
> It would look something like this
> 
> ```
> Error: File has an issue on line 3 at position 17
> Do you want to edit it? [y/N]: y
> Error: File has an issue on line 4 at position 9
> Do you want to edit it? [y/N]: n
> Error: Aborting due to wrong file format
> ```
> 
> Should I revert it? WDYT?
> 
> > I think that makes sense so we avoid problems with the format. In YAML there are many ways to write multilne string but we can force to use this one.
> 
> I missed this comment, sorry. I will check now how to enforce this (or should we just document this in the README.md?)
> 
> Edit: I'm sorry, I saw the instructions for how to create the changelogs are already present in the README.md [link](https://github.com/Bitergia/release-tools#whats-the-format-of-the-changelog-entries).

Just reviewed the PR. I think it makes sense to do it. Check my comments.

sduenas commented on release_tools/changelog.py:
```diff
@@ -211,13 +212,40 @@ def create_changelog_entry_content(title, category, author=None, issue=None,
+    is_valid = False
+
+    while not is_valid:
+
```
Can you remove this blank space?

sduenas reviewed (changes requested):
Looks better. Would you mind to add some tests? Also, it needs a changelog entry generate by this tool :D

vchrombie commented:
Hi @sduenas 
I have a small update on this.

I added the tests, but I was able to add only the valid format. I tried to add the test for an invalid format but wasn't successful, [code snippet](https://gist.github.com/vchrombie/4b61acf7f6c3482809c3420d99858493).

I tried to create a bad format dict, write it to file, and validate the contents of the file but I was stuck at [L33](https://gist.github.com/vchrombie/4b61acf7f6c3482809c3420d99858493#file-test_validate_entry-py-L33) as I can't write a wrong dict to the file. I might need some help to fix this. :confused:

sduenas commented:
> Hi @sduenas
> I have a small update on this.
> 
> I added the tests, but I was able to add only the valid format. I tried to add the test for an invalid format but wasn't successful, [code snippet](https://gist.github.com/vchrombie/4b61acf7f6c3482809c3420d99858493).
> 
> I might need some help to fix this. confused

You can mock `create_changelog_entry_content` so you can return any content you want. After, that the runner will try to validate the content.

sduenas commented on release_tools/changelog.py:
```diff
@@ -159,6 +159,18 @@ def changelog(title, category, dry_run, overwrite, editor):
     content = create_changelog_entry_content(title, category,
                                              run_editor=editor)
 
+    is_valid = False
+    while not is_valid:
+        is_valid = validate_changelog_content(content)
+
+        if not is_valid:
+            if not click.confirm("Do you want to edit it?"):
+                msg = "Aborting due to wrong format"
+                raise click.ClickException(msg)
+
+            # Allow the user to edit the content of the entry
+            content = click.edit(content)
+
```
I'd remove this to a different method so we keep the main function clean.

sduenas commented on release_tools/changelog.py:
```diff
@@ -211,13 +223,26 @@ def create_changelog_entry_content(title, category, author=None, issue=None,
+    except yaml.YAMLError as exc:
+        pm = exc.problem_mark
+        click.echo("Error: Invalid format on line {} at position {}".format(pm.line, pm.column))
+
```
Remove this blank line.

sduenas commented on releases/unreleased/validate-content-before-saving-with-the-changelog-tool.yml:
```diff
@@ -0,0 +1,12 @@
+author: Venu Vardhan Reddy Tekula <venu@bitergia.com>
+issue: 17
+notes: >
+    This code validates the format of the yml file and gives info
+    about the errors, if any. It gives an option to edit the file
+    incase it if has any mistakes.
+    
+    This feature is necessary to avoid any mistakes by the user
+    while editing the yml file using the changelog tool.
```
Instead of writing this as the comment on a commit, try to write it as if the reader were a user that doesn't know anything about the internals of the code. The user should understand what the new functionality is.

vchrombie replied:
Okay

sduenas commented on tests/test_changelog.py:
```diff
@@ -242,6 +242,21 @@ def test_entries_dir_not_created(self, mock_project):
             # Nothing was created in the directory
             self.assertEqual(len(os.listdir(fs)), 0)
 
+    def test_validate_good_entry(self):
```
As you are testing both cases, rename this function to `test_validate_entry`

sduenas replied:
We also need a test checking the whole process. Now that you have `validate_changelog_content` as a function, you can mock it and check what happens when it returns either `True` or `False`.

sduenas reviewed (changes requested):
In overall, it looks good but it needs some changes to have it perfect.

vchrombie commented:
I will resolve the remaining two comments by today. :+1:

sduenas commented on release_tools/changelog.py:
```diff
@@ -211,13 +212,42 @@ def create_changelog_entry_content(title, category, author=None, issue=None,
+        is_valid = validate_changelog_content(content)
+
+        if not is_valid:
+            if not click.confirm("Do you want to edit it?"):
```
Right now `confirm` default response is set to `no`. It should be set to `yes` so the user don't lose the information in case of clicking `intro` by mistake.
Also, I think we need to add extra information on the message to explain that, changes will be lost if the user doesn't create a valid file.

vchrombie replied:
Okay.

sduenas commented on release_tools/changelog.py:
```diff
@@ -211,13 +212,42 @@ def create_changelog_entry_content(title, category, author=None, issue=None,
+        if not is_valid:
+            if not click.confirm("Do you want to edit it?"):
+                msg = "Aborting due to wrong format"
+                raise click.ClickException(msg)
```
`confirm` can raise an `Abort` exception when `abort` parameter is set. Wouldn't it be better?

vchrombie replied:
Yeah, makes sense.

sduenas reviewed (changes requested):
LGTM, just minor comments.

sduenas reviewed (approved):
LGTM
